File: strategy/strategy.py
"""Base class for all wordle strategies"""

import typing as tp

import wordle.wordle as wordle
from strategy.py_strategy import Strategy


TStrategyFactory = tp.Callable[[], Strategy]


def tries_to_guess(strategy: Strategy, word: str) -> tp.List[str]:
    tries = []
    
    while True:
        guess = strategy.new_word()
        matches = wordle.get_matches(guess, word)
        strategy.record(guess, matches)
        tries.append(guess)

        if guess == word:
            break
    
    return tries



def build_stats(words: tp.Iterable[str], strategy_factory: TStrategyFactory) -> tp.List[int]:
    stats = [0]

    for word in words:
        tries = tries_to_guess(strategy_factory(), word)
        while len(stats) <= len(tries):
            stats.append(0)
        
        stats[len(tries)] += 1
    
    return stats

File: strategy/strategy.cpp
#include "strategy.hpp"

namespace wordle {

::std::vector<::std::string> TriesToSolve(Strategy *strategy,
                                          const ::std::string &word) {
  ::std::vector<::std::string> result;
  while (result.empty() || result.back() != word) {
    auto guess = strategy->NewWord();
    auto matches = GetMatches(guess, word);
    strategy->RecordMatches(guess, matches);
    result.push_back(guess);
  }
  return result;
}

::std::vector<int> BuildStats(const ::std::vector<::std::string> &words,
                              ::std::function<Strategy *()> factory) {
  std::vector<int> stats;

  for (auto word : words) {
    auto strategy = factory();
    auto tries = TriesToSolve(strategy, word);
    while (stats.size() <= tries.size()) stats.push_back(0);
    ++stats[tries.size()];
  }

  return stats;
}

}  // namespace wordle